internal/cephfs/controllerserver.go

```diff
@@ -298,3 +410,291 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
+					sid.FsSnapshotName, err)
+			}
+		}
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SizeBytes:      int64(info.BytesQuota),
+				SnapshotId:     sid.SnapshotID,
+				SourceVolumeId: req.GetSourceVolumeId(),
+				CreationTime:   sid.CreationTime,
+				ReadyToUse:     protected,
+			},
+		}, nil
+	}
+
+	// Reservation
+	sID, err := reserveSnap(ctx, parentVolOptions, vid.FsSubvolName, snapName, cr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer func() {
+		if err != nil {
+			errDefer := undoSnapReservation(ctx, parentVolOptions, *sID, snapName, cr)
+			if errDefer != nil {
+				klog.Warningf(util.Log(ctx, "failed undoing reservation of snapshot: %s (%s)"),
+					requestName, errDefer)
+			}
+		}
+	}()
+
+	snap := snapshotInfo{}
+	snap, err = doSnapshot(ctx, parentVolOptions, vid.FsSubvolName, sID.FsSnapshotName, cr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      int64(info.BytesQuota),
+			SnapshotId:     sID.SnapshotID,
+			SourceVolumeId: req.GetSourceVolumeId(),
+			CreationTime:   snap.CreationTime,
+			ReadyToUse:     true,
+		},
+	}, nil
+}
+
+func doSnapshot(ctx context.Context, volOpt *volumeOptions, subvolumeName, snapshotName string, cr *util.Credentials) (snapshotInfo, error) {
+	volID := volumeID(subvolumeName)
+	snapID := volumeID(snapshotName)
+	snap := snapshotInfo{}
+	err := createSnapshot(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s %v"), snapID, err)
+		return snap, err
+	}
+	defer func() {
+		if err != nil {
+			dErr := deleteSnapshot(ctx, volOpt, cr, snapID, volID)
+			if dErr != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapID, err)
+			}
+		}
+	}()
+	snap, err = getSnapshotInfo(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to get snapshot info %s %v"), snapID, err)
+		return snap, err
+	}
+	var t *timestamp.Timestamp
+	t, err = parseTime(ctx, snap.CreatedAt)
+	if err != nil {
+		return snap, err
+	}
+	snap.CreationTime = t
+	err = protectSnapshot(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to protect snapshot %s %v"), snapID, err)
+	}
+	return snap, err
+}
+
+func parseTime(ctx context.Context, createTime string) (*timestamp.Timestamp, error) {
+	tm := &timestamp.Timestamp{}
+	layout := "2006-01-02 15:04:05.000000"
+	// TODO currently parsing of timestamp to time.ANSIC generate from ceph fs is failng
+	var t time.Time
+	t, err := time.Parse(layout, createTime)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to parse time %s %v"), createTime, err)
+		return tm, err
+	}
+	tm, err = ptypes.TimestampProto(t)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to convert time %s %v"), createTime, err)
+		return tm, err
+	}
+	return tm, nil
+}
```

Comment:
does golint-ci/go-fmt not complain about missing empty lines between functions?!